File: app/src/main/java/me/pwcong/jpstart/component/activity/PuzzleActivity.java
package me.pwcong.jpstart.component.activity;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.AnimationSet;
import android.widget.Button;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import butterknife.BindView;
import me.pwcong.jpstart.R;
import me.pwcong.jpstart.conf.Constants;
import me.pwcong.jpstart.manager.SharedPreferenceManager;
import me.pwcong.jpstart.mvp.bean.JPItem;
import me.pwcong.jpstart.mvp.presenter.BasePresenter;
import me.pwcong.jpstart.mvp.presenter.PuzzleActivityPresenterImpl;
import me.pwcong.jpstart.mvp.view.BaseView;

/**
 * Created by Pwcong on 2016/10/24.
 */

public class PuzzleActivity extends BaseActivity implements BaseView.PuzzleActivityView, View.OnClickListener {

    public static final int TYPE_HIRAGANA_ROME = 0;
    public static final int TYPE_HIRAGANA_KATAKANA = 1;
    public static final int TYPE_KATAKANA_ROME = 2;

    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.tv_show)
    TextView mShowTextView;
    @BindView(R.id.tv_count)
    TextView mCountTextView;
    @BindView(R.id.tv_tips)
    TextView mTipsTextView;
    @BindView(R.id.btn_answer1)
    Button mButton1;
    @BindView(R.id.btn_answer2)
    Button mButton2;
    @BindView(R.id.btn_answer3)
    Button mButton3;
    @BindView(R.id.btn_answer4)
    Button mButton4;

    int type = 0;
    int count = 0;
    JPItem current;
    List<JPItem> items;

    BasePresenter.PuzzleActivityPresenter presenter;

    @Override
    protected int getViewId() {
        return R.layout.activity_puzzle;
    }

    @Override
    protected void initVariable(@Nullable Bundle savedInstanceState) {

        presenter = new PuzzleActivityPresenterImpl(this);

        initToolbar();
        initButton();
        initTextView();
    }

    private void initToolbar() {
        mToolbar.setTitle(R.string.whoami);
        mToolbar.setNavigationIcon(R.drawable.ic_chevron_left_white_24dp);
        setSupportActionBar(mToolbar);

        mToolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

    }

    private void initButton() {

        mButton1.setOnClickListener(this);
        mButton2.setOnClickListener(this);
        mButton3.setOnClickListener(this);
        mButton4.setOnClickListener(this);

    }

    private void initTextView() {
        mCountTextView.setText(String.valueOf(count));
    }

    @Override
    protected void doAction() {

        presenter.initPuzzleActivity();
    }

    @Override
    public void onClick(View v) {

        presenter.checkAnswerSelect(v.getId(), current, items);

    }

    @Override
    public void setData(JPItem current, List<JPItem> jams) {

        this.current = current;

        items = new ArrayList<>();
        items.add(current);
        items.addAll(jams);
        Collections.shuffle(items);

        switch (type) {

            case TYPE_HIRAGANA_ROME:

                mShowTextView.setText(this.current.getHiragana());

                mButton1.setText(this.items.get(0).getRome());
                mButton2.setText(this.items.get(1).getRome());
                mButton3.setText(this.items.get(2).getRome());
                mButton4.setText(this.items.get(3).getRome());

                break;
            case TYPE_HIRAGANA_KATAKANA:

                mShowTextView.setText(this.current.getHiragana());

                mButton1.setText(this.items.get(0).getKatakana());
                mButton2.setText(this.items.get(1).getKatakana());
                mButton3.setText(this.items.get(2).getKatakana());
                mButton4.setText(this.items.get(3).getKatakana());

                break;
            case TYPE_KATAKANA_ROME:

                mShowTextView.setText(this.current.getKatakana());

                mButton1.setText(this.items.get(0).getRome());
                mButton2.setText(this.items.get(1).getRome());
                mButton3.setText(this.items.get(2).getRome());
                mButton4.setText(this.items.get(3).getRome());

                break;
            default:
                break;

        }

    }

    @Override
    public void showSelectDialog(String[] selection) {

        new AlertDialog.Builder(this).setItems(selection, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {

                type = which;
                presenter.checkTypeSelect(which);

            }
        })
                .setCancelable(false)
                .create()
                .show();
    }

    @Override
    public void showResultDialog(int title, String msg, int icon,
            int pbt, DialogInterface.OnClickListener pbl,
            int nbt, DialogInterface.OnClickListener nbl) {

        new AlertDialog.Builder(this)
                .setCancelable(false)
                .setMessage(msg)
                .setTitle(title)
                .setPositiveButton(pbt, pbl)
                .setNegativeButton(nbt, nbl)
                .create()
                .show();

    }

    @Override
    public void showDialog(int icon, int title, String msg) {

        new AlertDialog.Builder(this)
                .setTitle(title)
                .setIcon(icon)
                .setMessage(msg)
                .create()
                .show();

    }

    @Override
    public void setTitle(String title) {
        mToolbar.setTitle(title);
    }

    @Override
    public void setTitle(int title) {
        mToolbar.setTitle(title);
    }

    @Override
    public void addCount() {
        count++;
        mCountTextView.setText(String.valueOf(count));
        mTipsTextView.setText(String.valueOf(count));
        mTipsTextView.startAnimation(getTipsAnimation());
    }

    @Override
    public void clearCount() {

        int hs = SharedPreferenceManager.getInstance().getInt(Constants.HIGHEST_SCORE, 0);
        if (count > hs)
            SharedPreferenceManager.getInstance().putInt(Constants.HIGHEST_SCORE, count);

        count = 0;
        mCountTextView.setText(String.valueOf(count));

    }

    @Override
    public void showMsg(int msg) {
        showSnackBar(mToolbar, msg);
    }

    @Override
    public void showMsg(String msg) {
        showSnackBar(mToolbar, msg);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        getMenuInflater().inflate(R.menu.activity_puzzle_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case R.id.menu_hiragana_rome:
                type = TYPE_HIRAGANA_ROME;
                presenter.checkTypeSelect(TYPE_HIRAGANA_ROME);
                break;
            case R.id.menu_hiragana_katakana:
                type = TYPE_HIRAGANA_KATAKANA;
                presenter.checkTypeSelect(TYPE_HIRAGANA_KATAKANA);
                break;
            case R.id.menu_katakana_rome:
                type = TYPE_KATAKANA_ROME;
                presenter.checkTypeSelect(TYPE_KATAKANA_ROME);
                break;

            default:

                presenter.checkMenuSelect(item.getItemId());

                break;

        }
        return true;
    }

    private Animation getTipsAnimation() {

        AnimationSet animationSet = new AnimationSet(true);
        animationSet.addAnimation(new AlphaAnimation(1, 0));
        animationSet.setDuration(1000);
        animationSet.setFillAfter(true);

        return animationSet;

    }
}

File: app/src/main/java/com/github/pwcong/jpstart/view/activity/PuzzleActivity.kt
package com.github.pwcong.jpstart.view.activity

import android.content.DialogInterface
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.AnimationSet
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import com.github.pwcong.jpstart.R
import com.github.pwcong.jpstart.constants.Constants
import com.github.pwcong.jpstart.databinding.ActivityPuzzleBinding
import com.github.pwcong.jpstart.manager.SharedPreferenceManager
import com.github.pwcong.jpstart.mvp.bean.JPItem
import com.github.pwcong.jpstart.mvp.presenter.BasePresenter
import com.github.pwcong.jpstart.mvp.presenter.impl.PuzzleActivityPresenterImpl
import com.github.pwcong.jpstart.mvp.view.BaseView.PuzzleActivityView
import java.util.Collections

class PuzzleActivity : BaseActivity<ActivityPuzzleBinding>(), PuzzleActivityView,
    View.OnClickListener {

    private lateinit var mToolbar: Toolbar

    private lateinit var mShowTextView: TextView

    private lateinit var mCountTextView: TextView

    private lateinit var mTipsTextView: TextView

    private lateinit var mButton1: Button

    private lateinit var mButton2: Button

    private lateinit var mButton3: Button

    private lateinit var mButton4: Button

    private var type = 0
    private var count: Int = 0

    private lateinit var current: JPItem
    private lateinit var items: MutableList<JPItem>

    private lateinit var presenter: BasePresenter.PuzzleActivityPresenter

    override fun initViewBinding(): ActivityPuzzleBinding {
        return ActivityPuzzleBinding.inflate(layoutInflater)
    }

    override fun init(savedInstanceState: Bundle?) {
        mToolbar = viewBinding.root.findViewById(R.id.toolbar)
        mShowTextView = viewBinding.tvShow
        mCountTextView = viewBinding.tvCount
        mTipsTextView = viewBinding.tvTips
        mButton1 = viewBinding.btnAnswer1
        mButton2 = viewBinding.btnAnswer2
        mButton3 = viewBinding.btnAnswer3
        mButton4 = viewBinding.btnAnswer4

        presenter = PuzzleActivityPresenterImpl(this)

        initToolbar()
        initButton()
        initTextView()

        presenter.initPuzzleActivity()
    }

    private fun initToolbar() {
        mToolbar.setTitle(R.string.whoami)
        mToolbar.setNavigationIcon(R.drawable.ic_chevron_left_white_24dp)
        setSupportActionBar(mToolbar)

        mToolbar.setNavigationOnClickListener { finish() }
    }

    private fun initButton() {
        mButton1.setOnClickListener(this)
        mButton2.setOnClickListener(this)
        mButton3.setOnClickListener(this)
        mButton4.setOnClickListener(this)
    }

    private fun initTextView() {
        mCountTextView.text = count.toString()
    }

    override fun onClick(v: View) {
        presenter.checkAnswerSelect(v.id, current, items)
    }

    override fun setData(current: JPItem, jams: List<JPItem>) {
        this.current = current

        items = mutableListOf()
        items.add(current)
        items.addAll(jams)
        Collections.shuffle(items)

        when (type) {
            TYPE_HIRAGANA_ROME -> {
                mShowTextView.text = this.current.hiragana

                mButton1.text = items.get(0).rome
                mButton2.text = items.get(1).rome
                mButton3.text = items.get(2).rome
                mButton4.text = items.get(3).rome
            }

            TYPE_HIRAGANA_KATAKANA -> {
                mShowTextView.text = this.current.hiragana

                mButton1.text = items.get(0).katakana
                mButton2.text = items.get(1).katakana
                mButton3.text = items.get(2).katakana
                mButton4.text = items.get(3).katakana
            }

            TYPE_KATAKANA_ROME -> {
                mShowTextView.text = this.current.katakana

                mButton1.text = items[0].rome
                mButton2.text = items[1].rome
                mButton3.text = items[2].rome
                mButton4.text = items[3].rome
            }

            else -> {}
        }
    }

    override fun showSelectDialog(selection: Array<String>) {
        AlertDialog.Builder(this).setItems(
            selection
        ) { _, which ->
            type = which
            presenter.checkTypeSelect(which)
        }
            .setCancelable(false)
            .create()
            .show()
    }

    override fun showResultDialog(
        title: Int, msg: String, icon: Int,
        pbt: Int, pbl: DialogInterface.OnClickListener?,
        nbt: Int, nbl: DialogInterface.OnClickListener?
    ) {
        AlertDialog.Builder(this)
            .setCancelable(false)
            .setMessage(msg)
            .setTitle(title)
            .setPositiveButton(pbt, pbl)
            .setNegativeButton(nbt, nbl)
            .create()
            .show()
    }

    override fun showDialog(icon: Int, title: Int, msg: String) {
        AlertDialog.Builder(this)
            .setTitle(title)
            .setIcon(icon)
            .setMessage(msg)
            .create()
            .show()
    }

    override fun setTitle(title: String) {
        mToolbar.setTitle(title)
    }

    override fun setTitle(title: Int) {
        mToolbar.setTitle(title)
    }

    override fun addCount() {
        count++
        mCountTextView.text = count.toString()
        mTipsTextView.text = count.toString()
        mTipsTextView.startAnimation(tipsAnimation)
    }

    override fun clearCount() {
        val hs = SharedPreferenceManager.getInstance().getInt(Constants.HIGHEST_SCORE, 0)
        if (count > hs) SharedPreferenceManager.getInstance().putInt(Constants.HIGHEST_SCORE, count)

        count = 0
        mCountTextView.text = count.toString()
    }

    override fun showMsg(msg: Int) {
        showSnackBar(mToolbar, msg)
    }

    override fun showMsg(msg: String) {
        showSnackBar(mToolbar, msg)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_puzzle_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_hiragana_rome -> {
                type = TYPE_HIRAGANA_ROME
                presenter.checkTypeSelect(TYPE_HIRAGANA_ROME)
            }

            R.id.menu_hiragana_katakana -> {
                type = TYPE_HIRAGANA_KATAKANA
                presenter.checkTypeSelect(TYPE_HIRAGANA_KATAKANA)
            }

            R.id.menu_katakana_rome -> {
                type = TYPE_KATAKANA_ROME
                presenter.checkTypeSelect(TYPE_KATAKANA_ROME)
            }

            else -> presenter.checkMenuSelect(item.itemId)

        }
        return true
    }

    private val tipsAnimation: Animation
        get() {
            val animationSet = AnimationSet(true)
            animationSet.addAnimation(AlphaAnimation(1f, 0f))
            animationSet.duration = 1000
            animationSet.fillAfter = true

            return animationSet
        }

    companion object {
        const val TYPE_HIRAGANA_ROME: Int = 0
        const val TYPE_HIRAGANA_KATAKANA: Int = 1
        const val TYPE_KATAKANA_ROME: Int = 2
    }
}
